pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,661 @@
+
+	createClusterWorkspace func(ctx context.Context, workspace logicalcluster.Name, cw *tenancyv1alpha1.ClusterWorkspace) error
+	getClusterWorkspace    func(ctx context.Context, workspace logicalcluster.Name, name string) (*tenancyv1alpha1.ClusterWorkspace, error)
+
+	getLocalClusterWorkspace      func(fullName logicalcluster.Name) (*tenancyv1alpha1.ClusterWorkspace, error)
+	searchLocalClusterRole        searchFunc
+	searchLocalClusterRoleBinding searchFunc
+	synced                        func() bool
+
+	homePrefix               logicalcluster.Name
+	bucketLevels, bucketSize int
+	creationDelay            int
+
+	// This is only for testing
+	mocks mockableFunctions
+}
+
+func (h *homeWorkspaceHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if !h.synced() {
+		h.apiHandler.ServeHTTP(rw, req)
+		return
+	}
+
+	ctx := req.Context()
+	var workspaceName logicalcluster.Name
+	if lcluster, err := request.ValidClusterFrom(ctx); err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	} else {
+		workspaceName = lcluster.Name
+	}
+	user, ok := request.UserFrom(ctx)
+	if !ok {
+		err := errors.New("No user in HomeWorkspaces filter !")
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	requestInfo, ok := request.RequestInfoFrom(ctx)
+	if !ok {
+		responsewriters.InternalError(rw, req, errors.New("No request Info"))
+		return
+	}
+
+	if workspaceName == tenancyv1alpha1.RootCluster &&
+		requestInfo.IsResourceRequest &&
+		requestInfo.Verb == "get" &&
+		requestInfo.APIGroup == tenancyv1beta1.SchemeGroupVersion.Group &&
+		requestInfo.Resource == "workspaces" &&
+		requestInfo.Name == "~" {
+		var homeWorkspace *tenancyv1beta1.Workspace
+		homeWorkspaceFullName := h.getHomeWorkspaceName(user.GetName())
+		homeClusterWorkspace, err := h.getLocalClusterWorkspace(homeWorkspaceFullName)
+		if err != nil {
+			responsewriters.InternalError(rw, req, err)
+			return
+		}
+		if homeClusterWorkspace != nil {
+			// The home workspace for the user exists, so let's return it to the user if he request to see its definition
+
+			homeWorkspace = &tenancyv1beta1.Workspace{}
+			projection.ProjectClusterWorkspaceToWorkspace(homeClusterWorkspace, homeWorkspace)
+		} else {
+			// Test if the user has the right to see his Home workspace even though it doesn't exists
+			// => test the get verb on the clusterworkspaces/workspace subresource named ~ in the root workspace.
+
+			attributes := homeWorkspaceAuthorizerAttributes(user, "get")
+			decision, reason, err := h.authz.Authorize(ctx, attributes)
+			if err != nil {
+				responsewriters.InternalError(rw, req, err)
+				return
+			}
+			if decision != authorizer.DecisionAllow {
+				responsewriters.Forbidden(ctx, attributes, rw, req, reason, homeWorkspaceCodecs)
+				return
+			}
+
+			parent, name := homeWorkspaceFullName.Split()
+			homeWorkspaceURL := &url.URL{Scheme: "https", Host: req.Host, Path: homeWorkspaceFullName.Path()}
```

Comment:
this is wrong. `req.Host` is the shard if I am not mistaken. This must be the front-proxy.

Reply:
Fixed in commit b631fc4caa9b4c4da8883dff9b282836eb6dcf25